### PR TITLE
Add when conditions to commonly used processors

### LINKED ISSUE #2613: Add conditional support to commonly used processors
**Is your feature request related to a problem? Please describe.**
As a user of processors, I would like to only run processors conditionally based on the Events using the Data Prepper expression syntax

**Describe the solution you'd like**
Add conditional support to the following processors

```
* grok
* parse_json
* mutate event processors
* mutate string processors
```

Following from the existing naming scheme for processors that support this conditional parameter (`aggregate` processor has `aggregate_when`, drop processor has `drop_when`, the conditional statement parameters for these processors will follow the same schema (i.e. `grok_when` (for grok), `parse_when` (for parse_json), `add_when` (for add_entries), and so on).

The alternative to this is using a common parameter for all the conditional statement parameters, that being `when`. The benefit of this would be to eliminate breaking changes being made in the case that we add support in the future for conditional statements being at the processor level with a common parameter name of `when`. 

**Describe alternatives you've considered (Optional)**
Adding support through data-prepper-core to support filtering events based on a `when` parameter that can automatically be used for all processors. This could still be something we do in the future, but as it is a core change, it comes with extra complexity

**Additional context**

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
This change adds "when" conditional support to the following processors
* grok (`grok_when`)
* parse_json (`parse_when`)
* add_entries (`add_when` on each entry)
* convert_entry_type (`convert_when`)
* copy_value (`copy_when` on each entry)
* delete_entries (`delete_when`)
* list_to_map (`list_to_map_when`)
* rename_keys (`rename_when` on each entry)
* split_string (`split_when` on each entry)
* substitute_string (`substitute_when` on each entry)

This PR also fixes a bug that occurred when the `parse_json` processor is used on a key that does not exist in the Event, resulting in an an exception that results in Events being dropped
 
### Issues Resolved
closes #2613 
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java [hunk 1]
```diff
@@ -43,17 +46,24 @@ public boolean getOverwriteIfKeyExists() {
             return overwriteIfKeyExists;
         }
 
+        public String getAddWhen() { return addWhen; }
```
Comment (oeyh) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java:49]: Should we allow another `add_when` on the same level as `entries`, which has the benefit of combining multiple entries to add under one condition? Same for other processors with `entries`.

**Review by oeyh — COMMENTED** [review]
Looks great! One comment below. Also, can we update the doc as a follow-up?

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java (hunk 1 above)
Comment (graytaylor0) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java:49]: We can but I think in the future we will have a global processor "when" option which will be used in that place, so I don't think it's necessary at this time.

**Review by oeyh — APPROVED** [review]

**Comment by graytaylor0** [comment]
> Looks great! One comment below. Also, can we update the doc as a follow-up?

Good call will update the docs in a follow up PR

**Comment by dlvenable** [comment]
@graytaylor0 , More important that the documentation in this repository is the user documentation on opensearch.org. Please create an issue in https://github.com/opensearch-project/documentation-website requesting documentation updates for this feature.

**Review by dlvenable — APPROVED** [review]

**Pull request merged**